### PR TITLE
Dont block UI if .updateWith() not called (closes #589)

### LINKED ISSUE #589: .updateWith() should only be callable if event was canceled
The way the spec is currently written, the payment sheet of a browser can lock up waiting on `ev.updateWith()` to be called. 

What actually needs to happen is: 

  1. That the developer needs to cancel the event somehow and then call `ev.updateWith()` - maybe .updateWith() cancels the event automatically? (I like this, because it means existing apps can continue to work without devs needing to make any changes to their existing code! And browsers need only make a small change.)
  1. the browser needs to check if the event was cancelled. If yes, wait for `.updateWith()`. If no, reuse the existing data (or present whatever data the user has changed - like a new address or shipping option). Payment sheets already do this when changing credit cards, email, phone number, for instance (where no event is fired at the page). 

So, Proposal: 

* Calling `updateWith()` sets the event's cancel flag. 
* In the sheet:
   * If the event was not cancelled, then just update reusing the existing data + whatever new data the user provide/changed.
* Otherwise, wait for `.updateWith()`'s promise to settle. 

Pre-empting discussion: checking if an event listener/handler is set on the payment request would [violate the DOM spec](https://dom.spec.whatwg.org/#observing-event-listeners) (because it would be a side-effect). That would be very bad.

### PR DESCRIPTION
**Pull request by marcoscaceres**
Tests:
w3c/web-platform-tests#7288

***
[Preview](https://s3.amazonaws.com/pr-preview/w3c/payment-request/cancel_flag.html) | [Diff](https://s3.amazonaws.com/pr-preview/w3c/payment-request/2cefb58...0e8e780.html)

### CODE REVIEW

File: index.html
Comment (domenic) [inline index.html]: So this would happen if someone calls updateWith(p), p settles, and then someone calls updateWith(q), right?

Whereas the next line would happen if someone calls updateWith(p), and before p settles, someone calls updateWith(q)?

If I got that right, then I'd suggest adding quick examples or notes inside these steps re-stating the above.

File: index.html
Comment (domenic) [inline index.html]: This doesn't work. Fire an event returns a boolean, not an event.

You should instead:

- Let _event_ be the result of [creating an event](https://dom.spec.whatwg.org/#concept-event-create) using Event.
- Initialize _event_'s type attribute to _name_
- [Dispatch](https://dom.spec.whatwg.org/#concept-event-dispatch) _event_ at _target_.

File: index.html
Comment (domenic) [inline index.html]: ", then re-enable the relevant parts of user interface", I think.

Maybe this should be a series of nested steps under the "If ... is true".

**Review by domenic — CHANGES_REQUESTED** [review]

File: index.html
Comment (marcoscaceres) [reply index.html]: > This doesn't work. Fire an event returns a boolean, not an event.

Argh, I should have remembered that. I've used that boolean before elsewhere.

File: index.html
Comment (marcoscaceres) [reply index.html]: `.updateWith()` controls re-enabling the user interface, in step 14. It might be redundant to re-state it here.

Comment (domenic) [reply index.html]: Ah ok, in that case, agreed.

File: index.html
Comment (marcoscaceres) [reply index.html]: > So this would happen if someone calls updateWith(p), p settles, and then someone calls updateWith(q), right?

I was thinking of this case: 

```JS
request.onaddresschange  = ev => { 
   Promise.resolve({...details})
     // next tick, so [[didUpdate]] is now true. Throws InvalidStateError.
     .then(ev.updateWith.bind(ev));
}
```

>Whereas the next line would happen if someone calls updateWith(p), and before p settles, someone calls updateWith(q)?

Correct. It also prevents someone holding a reference to event and calling `.updateWith(p)` after `p` has settled. The problem in the current spec is that [[\waitForUpdate]] and request.[[updating]] are both reset to `false` in step 14 of `updateWith()` - so theoretically, a developer could have just waited like this:  

```JS
request.onaddresschange  = async ev => {
   const p =  Promise.resolve({...details});
   ev.updateWith(p);
   await p;  // [[waitForUpdate]] and request.[[updating]] are both reset to `false`! 
   // [[didUpdate]] is true, this now throws
   const q = Promise.resolve({...details});
   ev.updateWith(q);
}
```

Comment (marcoscaceres) [reply index.html]: Fixed first case above...

Comment (marcoscaceres) [reply index.html]: Ok, added example above.

File: index.html
Comment (domenic) [inline index.html:2405]: I'm just noticing that this repeats text from slightly above. Let's delete the above text, as it feels pretty out of place anyway.

File: index.html
Comment (domenic) [inline index.html]: No `async` here

File: index.html
Comment (domenic) [inline index.html]: "of course" is a bit aggressive and implies the reader understands what's going on. You shouldn't really say "of course" while teaching someone something like this, IMO.

File: index.html
Comment (domenic) [inline index.html]: Let's do `await getNewDetails(oldDetails)` to mirror the below.

File: index.html
Comment (domenic) [inline index.html]: Backticks don't work in example titles. Respec bug? Or just remove it for now?

Same below.

File: index.html
Comment (domenic) [inline index.html]: In addition to the backticks, updateWith() is misspelled

File: index.html
Comment (domenic) [inline index.html]: [[didUpdate]] and [[waitForUpdate]] together prevent reuse

File: index.html
Comment (domenic) [inline index.html]: no `async`

File: index.html
Comment (domenic) [inline index.html]: To distinguish this from the last example, we should say something like "await goes to next tick, and updateWith() was not called before that, so [[didUpdate]] gets set to true".

Although that calls into question whether didUpdate is the right name, since nothing actually updated.

File: index.html
Comment (domenic) [inline index.html]: To distinguish this from the first example, we can say "the update has now completed, so [[didUpdate]] gets set to true; this throws "InvalidStateError".

File: index.html
Comment (domenic) [inline index.html]: Ideally link type, or at least `<code>`ify it.

File: index.html
Comment (domenic) [inline index.html]: event.updateWith()'s detailsPromise is pretty terrible. Can we store it in an internal slot?

Actually, what is the point of this "wait" step? Nothing happens after it. I think it should just be removed.

**Comment by zkoch** [comment]
Is there a TL;DR for what we're trying to accomplish here? It's not obvious to me. :)

**Comment by marcoscaceres** [comment]
@zkoch, tldr:

 1. Don't block the payment sheet if the developer doesn't call `.updateWith()`. 
 1. Don't allow developers to keep recycling the event (e.g., keep a reference to a shipping change event, and then use it to update the the payment sheet when an option change event happens - see code below).  

===

Note to self, add this test: 
```js
var methods = [{ supportedMethods: ["basic-card"] }];
var details = {
  total: { label: "", amount: { currency: "USD", value: "1" } },
};
var shippingOptions = [
  {
    id: "standard",
    label: "🚛 Ground Shipping (2 days)",
    amount: { currency: "USD", value: "5.00" },
    selected: true,
  },
  {
    id: "drone",
    label: "🚀 Drone Express (2 hours)",
    amount: { currency: "USD", value: "25.00" },
  },
];
details.shippingOptions = shippingOptions;
var r = new PaymentRequest(methods, details, { requestShipping: true });
r.addEventListener(
  "shippingaddresschange",
  e => {
    ev = e;
    ev.updateWith(details);
    var listener = ev2 => {
      var newDetails = { ...details };
      newDetails.total.label = "FAIL" + Math.random();
      ev.updateWith(newDetails);
    };
    r.addEventListener("shippingoptionchange", listener);
    r.addEventListener("shippingaddresschange", listener);
  },
  { once: true }
);
r.show();
```

File: index.html
Comment (marcoscaceres) [reply index.html]: I know... "didUpdate" is not a great name... I wanted something that denoted "consumed" or "expired" or "this can't be used again". I'm open to suggestions.

Comment (dlongley) [reply index.html]: Could flip it and go with "canUpdate" or "updatePermitted" and turn it off after the opportunity to update is lost or has already been taken.

File: index.html
Comment (marcoscaceres) [reply index.html]: Agree. I wasn't sure if it would be clear. But if one follows the updateWith() path, it provides all the steps there.

File: index.html
Comment (marcoscaceres) [reply index.html]: I plan to add support for this in ReSpec, so will leave it for now and we will get this for free soon.

File: index.html
Comment (marcoscaceres) [reply index.html]: > Although that calls into question whether didUpdate is the right name, since nothing actually updated.

Rethinking this: it actually did update... but instead of using the developer's `details`, the sheet used the details that were provided by the payment sheet itself. So maybe `updatedUI` might be better.
 
@dlongley, I'd prefer not to flip it. I don't want to rewrite it :(

File: index.html
Comment (marcoscaceres) [reply index.html:2405]: deleted the one above.

File: index.html
Comment (domenic) [inline index.html]: Can this be code-ified *and* linked? :)

**Review by domenic — APPROVED** [review]
LGTM with one remaining nit; please don't wait on another round of review before merging :).

**Comment by adrianba** [comment]
I think the current spec already blocks this. The waitForUpdate slot is essentially equivalent to the new slot introduced. To avoid re-use of events, we should remove step 14, part 1, which sets waitForUpdate back to false.

**Comment by marcoscaceres** [comment]
> I think the current spec already blocks this. The waitForUpdate slot is essentially equivalent to the new slot introduced. To avoid re-use of events, we should remove step 14, part 1, which sets waitForUpdate back to false.

Nice!

**Comment by marcoscaceres** [comment]
@adrianba, I removed the new slot as you suggested. Adding appropriate tests and see if this works. Will carry over domenic's R+ if it works ok.

**Comment by marcoscaceres** [comment]
Rebased.

**Comment by marcoscaceres** [comment]
Moved test requirement to WPT repo.
https://github.com/w3c/web-platform-tests/issues/7288

I've got this mostly tested locally. Tests will hopefully land in the next few days. 

I feel confident merging this now having tested it.

**Pull request merged**
